### PR TITLE
feat: lw6 blogpost

### PR DESCRIPTION
**Pull request by J0**
## What kind of change does this PR introduce?

Blogpost for LW6
Tasks required to complete before merging
- [x] Include flow diagram from design team
- [x] Proofread + generate preview locally
- [x] Include users who have given consent in the thanks section at the bottom
- [x] Update Metadata at the top of post(image, thumb, tags + check if @hf already has metadata in the author list)

## What is the current behavior?

Please link any relevant issues here.

## What is the new behavior?

Feel free to include screenshots if it includes visual changes.

## Additional context

Add any other context or screenshots.

### CODE REVIEW

File: apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+ toc_depth: 3
+ ---
+
+
+# Multi-factor Authentication via Row Level Security Enforcement
+
+
+## Background
+
+Today, we’re releasing Multi Factor Authentication for everyone. We are also excited to commence the dogfooding of SAML support via the introduction of Single Sign On (SSO) on our dashboard - you can contact us at [growth@supabase.com](mailto:growth@supabase.com) if you want to enable this in your Enterprise or Pro plan.
```
Comment (hf) [inline apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx:20]: Link to `https://docs.supabase.com/guides/platform/sso/enable-sso` here where appropriate.

File: apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+## What is MFA?
+
+Multi-factor authentication (MFA), sometimes called two-factor authentication (2FA), adds an additional layer of security to your application by letting you verify users’ identity through extra steps. This typically consists of something you know, like a password, and something you have, like an authenticator application. We built MFA in response to customer requests - developers wanted enhanced security- be it for compliance, client requirements, or simply for peace of mind. As such, we started by building MFA support for Time-Based One Time Passwords (TOTP).
```
Comment (hf) [inline apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx:24]:
```suggestion
Multi-factor authentication (MFA), sometimes called two-factor authentication (2FA), adds an additional layer of security to your application by letting you verify users’ identity through extra steps. This typically consists of something you know, like a password, and something you have, like an authenticator application. We built MFA in response to customer requests - developers wanted enhanced security - be it for compliance, client requirements, or simply for peace of mind. As such, we started by building MFA support for Time-Based One Time Passwords (TOTP).
```

File: apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+
+## An Overview of TOTP
+
+[TOTP](https://www.rfc-editor.org/rfc/rfc6238) works by generating a unique, one-time password that is valid for a limited amount of time, usually 30 seconds or less. This password is generated using a shared secret key that is known only to the device and Supabase Auth, along with the current time. To exchange the shared secret key, a user scans a QR code generated by the server in order to establish a connection.  The QR code can be represented by a URI which conforms to the [Google Authenticator Key URI format](https://github.com/google/google-authenticator/wiki/Key-Uri-Format):
```
Comment (hf) [inline apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx:29]:
```suggestion
[TOTP](https://www.rfc-editor.org/rfc/rfc6238) works by generating a unique, one-time password that is valid for a limited amount of time, usually 30 seconds or less. This password is generated using a shared secret key that is known only to the device and Supabase Auth, along with the current time. To exchange the shared secret key, a user scans a QR code generated by the server in order to establish a connection. The QR code can be represented by a URI which conforms to the [Google Authenticator Key URI format](https://github.com/google/google-authenticator/wiki/Key-Uri-Format):
```

File: apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+otpauth://totp/supabase.io:j@supacats.io?algorithm=SHA1&digits=6&issuer=supabase.io&period=30&secret=BFSXQHFB2BGAZIOQWCDBJUF7B54A52JQ
+```
+
+The first portion of  `otpauth://totp/supabase.io` describes the protocol and issuer while `j@supacats.io` refers to the user. The remaining parameters refer to specifics around OTP generation. In this case, the OTP code is generated using a SHA1 hash of the secret combined with the timestamp and the OTP code is valid for `30s`
+
+In the event that the user faces difficulties entering a QR code the user can also opt to manually type the secret into the authenticator device.
+
+## MFA Flows: Enrolment and Verification
+
+An MFA flow can be broken into two key steps: Enrollment and Verification. During the *Enrollment* process Supabase Auth exchanges secrets with the user’s authenticator application. During the *Verification* process, the device makes use of the timestamp together with the secret to produce a six digit code that the server can verify.
```
Comment (hf) [inline apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx:41]:
```suggestion
An MFA flow can be broken into two key steps: Enrollment and Verification. During the *Enrollment* process Supabase Auth exchanges a randomly generated secret with the user’s authenticator application. During the *Verification* process, the device makes use of the timestamp together with the secret to produce a six digit code that the server can verify.
```

File: apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+You might wonder: why the need for the challenge step? The challenge step is designed to create an interval between MFA initiation and the action of making a verification. This is useful in cases like Yubikey authentication where a user might need to request a challenge before placing their finger on the device.
+
+```mermaid
+graph TD
+    Z[Start] --> A
+    A[Enroll Factor] --> B(Challenge Factor) -->  K(Verify Factor)
+    K -->|Correct Code| D[Factor Verified, Logged in ]
+    K -->|Incorrect Code/Error| Z
+
+```
```
Comment (hf) [inline apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx:56]: Can we verify that `mermaid` is supported _before_ we publish the post?

File: apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+**Verification**
+
+On subsequent logins attempts, redirect a user to an MFA verification after they have completed the conventional sign in process. On the verification page, wait for the user to enter the six digit OTP code from the authenticator application and then call  the`/challenge` endpoint followed by the `/verify` endpoint. If a correct code is submitted, a JWT will be created with a few additional fields.
+
+```mermaid
```
Comment (hf) [inline apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx:62]: Ditto.

File: apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+```
+
+## Enforcement via Row Level Security
+
+Adding MFA to your project and not enforcing it is like buying an expensive door and never locking it. We love Postgres RLS at Supabase and we believe it plays a key role in facilitating custom checks for a spread of use cases. To support RLS integration, JWTs issued by Supabase Auth now contain two additional pieces of information:
+
+1. An Authenticator Assurance Level (AAL) claim. Use this to quickly identify the level of identity checks the user has performed. `aal1` is reserved for conventional sign in, while `aal2` is issued only after the user has verified with an additional factor.
+
+2. An Authenticator Method Reference (AMR) claim. Use this to identify all of the authentication methods used by the user. This is also useful if you wish to implement step-up login scenarios.
```
Comment (hf) [inline apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx:77]:
```suggestion
1. An Authenticator Assurance Level (AAL) claim. Use this to quickly identify the level of identity checks the user has performed. `aal1` is reserved for conventional sign in, while `aal2` is issued only after the user has verified with an additional factor.
2. An Authenticator Method Reference (AMR) claim. Use this to identify all of the authentication methods used by the user. This is also useful if you wish to implement step-up login scenarios.
```

File: apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+      "timestamp": 1670924394
+    },
+    {
+      "method": "totp",
+      "timestamp": 1670925771
+    }
+  ],
+  "session_id": "9db34b97-a0e3-4eaf-810c-00b21b844cd0"
+}
+```
```
Comment (hf) [inline apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx:107]: May be better to just highlight the important bits?

```suggestion
```jsx
{
  "sub": "8802c1d6-c555-46e3-aacd-b61198b058d9",
  "email": "j0@supacats.io",
  "aud": "authenticated",
  "exp": 1670929371,
  "aal": "aal2",
  "amr": [
    {
      "method": "password",
      "timestamp": 1670924394
    },
    {
      "method": "totp",
      "timestamp": 1670925771
    }
  ],
  // ...
}
```
```

**Comment by hf** [comment]
Maybe link to the guide somewhere for an in-depth explanation of how to add MFA to your app?

File: apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+
+```jsx
+otpauth://totp/supabase.io:j@supacats.io?algorithm=SHA1&digits=6&issuer=supabase.io&period=30&secret=BFSXQHFB2BGAZIOQWCDBJUF7B54A52JQ
+```
+
+The first portion of  `otpauth://totp/supabase.io` describes the protocol and issuer while `j@supacats.io` refers to the user. The remaining parameters refer to specifics around OTP generation. In this case, the OTP code is generated using a SHA1 hash of the secret combined with the timestamp and the OTP code is valid for `30s`
+
+In the event that the user faces difficulties entering a QR code the user can also opt to manually type the secret into the authenticator device.
+
+## MFA Flows: Enrolment and Verification
```
Comment (kangmingtay) [inline apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx:41]:
```suggestion
## MFA Flows: Enrollment and Verification
```

File: apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx [hunk 10]
```diff
@@ -0,0 +1,141 @@
+---
+ title: 'Multi-factor Authentication via Row Level Security Enforcement'
+ description: MFA Auth with enforcement via RLS
+ author: joel_lee,kangmingtay,stojan
+ image: lw6-auth/thumbnail.jpg
+ thumb: lw6-auth/thumbnail.jpg
```
Comment (kangmingtay) [inline apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx:6]: maybe rename `thumb.jpg` to something clearer like `mfa-banner.jpg` ?
```suggestion
 image: lw6-auth/thumb.jpg
 thumb: lw6-auth/thumb.jpg
```

File: apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+
+Be mindful of your user’s preference, though. If a user has enabled MFA, they are expecting a higher level of security for their account. Consequently, we recommend that developers enforce MFA across all operations if a user has MFA enabled.
+
+
+## What's Next
+
+For starters, we are looking to support WebAuthn and FIDO2 compliant devices such as Yubikeys. Further, we will also be looking to facilitate easier configuration of MFA via email triggers on key actions.  If you have MFA requirements which are not covered here feel free to write to us at support[at]supabase.io .
+
+We are grateful to our early MFA users for the support and feedback provided throughout this period. In particular, we would like to thank [Fabian Beer](https://madebyfabian.com), [Cogram](https://www.cogram.com), and [Happl](https://happl.com) whose detailed feedback helped to shape our implementation. We would also like to specially thank the community behind [otp](https://github.com/pquerna/otp) and [svgo](https://github.com/ajstarks/svgo) libraries - their work is indispensable to this implementation.
```
Comment (kangmingtay) [inline apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx:141]:
```suggestion
We are grateful to our early MFA users for the support and feedback provided throughout this period. In particular, we would like to thank [Fabian Beer](https://madebyfabian.com), [Cogram](https://www.cogram.com), and [Happl](https://happl.com) whose detailed feedback helped to shape our implementation. We would also like to specially thank the community behind the [pquerna/otp](https://github.com/pquerna/otp) and [ajstarks/svgo](https://github.com/ajstarks/svgo) libraries - their work is indispensable to this implementation.
```

File: apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+toc_depth: 3
+---
+
+
+# Multi-factor Authentication via Row Level Security Enforcement
+
+
+## Background
+
+Today, we’re releasing Multi Factor Authentication for everyone. We are also excited to commence the dogfooding of SAML support via the introduction of Single Sign On (SSO) on our dashboard - you can contact us at [growth@supabase.com](mailto:growth@supabase.com) if you want to enable this in your Enterprise or Pro plan.
```
Comment (J0) [inline apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx:22]: We can merge this once the guide goes in.

```suggestion
[enable this in your Enterprise or Pro plan](https://docs.supabase.com/guides/platform/sso/enable-sso).
```

File: apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx (hunk 5 above)
Comment (J0) [reply apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx:59]: Looks like it isn't :( Will go with adding images unless the mermaid-mdx works fine and frontend approves of it

File: apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+*Enforce MFA for selected users*
+
+Note that both RLS policies are restrictive. By default, overlapping policies in PostgreSQL are permissive rather than restrictive. This means that RLS policies are combined with an `OR` clause and only one policy needs to pass in order for a row to be operated on. Therefore, we set RLS policies as restrictive to enforce the checks from multiple policies.
+
+Be mindful of your user’s preference, though. If a user has enabled MFA, they are expecting a higher level of security for their account. Consequently, we recommend that developers enforce MFA across all operations if a user has MFA enabled. You can check out [our MFA guide](https://supabase.com/docs/guides/auth/auth-mfa) for more details about MFA enforcement.
+
+
+## What's Next
+
+For starters, we are looking to support WebAuthn and FIDO2 compliant devices such as Yubikeys. Further, we will also be looking to facilitate easier configuration of MFA via email triggers on key actions.  If you have MFA requirements which are not covered here feel free to write to us at support[at]supabase.io .
```
Comment (J0) [inline apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx:131]:
```suggestion
For starters, we are looking to support WebAuthn and FIDO2 compliant devices such as Yubikeys. We also hope to allow users to receive email notifications when selected MFA actions are triggered. If you have MFA requirements which are not covered here feel free to write to us at support[at]supabase.io .
```

File: apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+  - launch-week
+  - auth
+  - mfa
+  - saml
+date: '2022-12-14'
+toc_depth: 3
+---
+
+
+# Multi-factor Authentication via Row Level Security Enforcement
```
Comment (J0) [inline apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx:17]:
```suggestion
```

File: apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+
+
+## Background
```
Comment (J0) [inline apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx:20]:
```suggestion
```

**Review by RamiroND — APPROVED** [review]

**Pull request merged**